### DISCUSSION TITLE
Are there any ways to sort or filter abundant validation errors to make it more clear what needs to be fixed?

### DISCUSSION THREAD

**Discussion by allenlawrence94** [pydantic/pydantic#5234] [Question] [answered]
I'm working on an API that uses lots of different pydantic models that are up to ~5 layers deep. I'll give a toy example of my issue:

Say my request body is 99% valid, but one deeply nested attribute `Model.SubModel.SubModel1.SubModel2.foo` is the wrong type. Now, the error message that pydantic returns says a lot more than just "foo is the wrong type"; it also tells me why the request body doesn't fit every other combination of models and submodels, so my API returns like 20 errors. I understand that this behavior is a feature, not a bug, but users of my API tend to struggle to parse these errors and figure out what went wrong with their request body.

Are there any good ways to sort or filter pydantic's validation errors to make it more clear what needs to be fixed? Maybe it's as simple as selecting for the deepest error(s). Does pydantic (or any other package) support anything like this? Or maybe I could refactor my models in some way to avoid the error bloat? Any ideas would be much appreciated.

**Comment by dmontagu** [comment, marked as answer]
Are you using lots of unions (of BaseModels)? In that case, you may get some benefits out of discriminated unions by specifying a discriminator attribute. I'd be happy to show an example of this if the docs aren't enough for you: https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions

If not, it would be helpful if you could share some code and explain how you'd like it to work. (This is part of the user experience I would be keen to improve; I do agree that, as a user, it can be difficult to unwind the overly-thorough error messages.)

**Reply by allenlawrence94** [reply]
I'm using a lot of unions! Discriminated unions seem super promising. Thanks so much for the quick reply!

**Reply by allenlawrence94** [reply]
I also still may try implementing an exception handler that sorts by the depth of the error. Does this intuition make sense to you?

**Reply by dmontagu** [reply]
> I also still may try implementing an exception handler that sorts by the depth of the error. Does this intuition make sense to you?

This makes sense, if you succeed, I'd be interested to see the output. I think it would require some changes to work with v2 but if the results are compelling I could see it being a good idea for incorporation as standard behavior.

**Reply by allenlawrence94** [reply]
Is there a way to make a discriminator optional? Maybe this is very niche, but as I'm now trying to migrate an API from using regular unions to discriminated unions, it would be very useful to have the discriminator field be optional for backwards compatibility.

**Reply by dmontagu** [reply]
Can you provide more context? Some example code showing what you want to do that doesn't work, or more detail about what sort of error you want that you aren't getting would be helpful.

**Reply by allenlawrence94** [reply]
I'd like to have something like this:
```python
from pydantic import BaseModel, Field
from typing import Union, Literal, Optional


class A(BaseModel):
    typ: Optional[Literal["a"]]
    foo: int


class B(BaseModel):
    typ: Optional[Literal["b"]]
    bar: int


class C(BaseModel):
    baz: Union[A, B] = Field(..., discriminator="typ")


# this doesn't work
jazz = C(baz={"bar": 2})
assert isinstance(jazz, B)
```

My hope is to preserve the discriminator-less union behavior when the discriminator isn't provided, but reap the benefits of the discriminator when it is provided.

I'm figuring this isn't an option, and I should just build a parallel schema that uses discriminators... unless you have a better idea.

**Reply by dmontagu** [reply]
I think this will be easier to do in v2. In v1... you have some options but I'm struggling to come up with something without too much code.

If you don't mind setting a default value besides None on `typ`, one possibility might be:

```python
from pydantic import BaseModel, Field, validator, parse_obj_as
from typing import Union, Literal, Optional


class A(BaseModel):
    typ: Optional[Literal["a"]] = 'a'
    foo: int


class B(BaseModel):
    typ: Optional[Literal["b"]] = 'b'
    bar: int


class C(BaseModel):
    baz: Union[A, B] = Field(..., discriminator="typ")

    @validator('baz', pre=True)
    def validate_baz(cls, values):
        if 'typ' in values:
            return values
        parsed = parse_obj_as(Union[A, B], values)
        values['typ'] = parsed.__fields__['typ'].default
        return values
```

While this requires a bit of case-specific code, it's also fairly generic. (You could get rid of the need for default by using a ClassVar on each type that holds the discriminator value.)

Is that helpful?

**Reply by allenlawrence94** [reply]
Yes this is much better than what I would have done! I'll give this a go. Thanks again for all your help!

**Reply by allenlawrence94** [reply]
I ended up landing on this class decorator that seems to do the trick:
```python
import typing as t
from pydantic import BaseModel, parse_obj_as
from pydantic.fields import Validator


def optional_discriminators(unions: t.List[str]):
    def dec(model: t.Type[BaseModel]):
        for k in unions:
            field = model.__fields__[k]
            discriminator_lookup = {v.type_: k for k, v in field.sub_fields_mapping.items()}

            def handle_missing_discriminator(cls, values):
                if field.discriminator_key not in values:
                    parsed = parse_obj_as(t.Union[tuple(f.type_ for f in field.sub_fields)], values)
                    values[field.discriminator_key] = discriminator_lookup[type(parsed)]
                return values

            field.class_validators[f"handle_missing_{field.discriminator_key}"] = Validator(
                handle_missing_discriminator, pre=True
            )
            field.populate_validators()
        return model

    return dec
```

Used like so:
```python
@optional_discriminators(["baz"])
class C(BaseModel):
    baz: Union[A, B] = Field(..., discriminator="typ")
```

Posting here as I figure this could be useful for someone else

**Reply by allenlawrence94** [reply]
This doesn't require default values on the `Literal` fields
